### ISSUE TITLE
Promises in fetch

### ISSUE THREAD

**Issue by soyjavi** [spencermountain/wtf_wikipedia#32]
A month ago I started creating my own wiki/parser but as you know it's a big pain in the ass. So I decided colaborate with another library, yesterday I found this one and I think is a great start. My idea is helping you with this, but I think we have different code-styles, and before start send you PRs I wanna tell you my thoughts:
- Why we need more than one method? Now we have `plaintext`, `from_api` and `parse`. IMHO this library must be a unique stateless method:

``` js
import wikipediaWTF from `wikipedia-wtf`;

wikipedia(sentence {string}, options {language...})
```
- Why we dont use `Promises` like?Better if we use something like:

``` js
wikipedia.from(sentence, options)
  .then(value => ...)
  .catch(error => ...)
```
- If it's a promise we can use the new standard `async/await`:

```
const knowledge = await wikipedia(sentence, options);
```
- IMHO is better approach auto-parse the markup to a JavaScript Object
- Now we have `infobox` and `infobox_template`, why we dont create a unique attribute?**
- And the last one (for now), why we use `grunt`?? Maybe we have to move to `babel` + `mocha` via NPM tasks (zero-dependencies)

I think this library is a great start, but we have to create an consistent DSL for a better experience using it.

best,

javi

**Comment by spencermountain** [comment]
hi Javi, thanks for the kind words. yes, this library needs a lot of attention. I welcome all the help you have.
This library is, as far as I remember, synchronous - straight-up text methods - so promises aren't necessary, in my opinion.  You may be interested in the work @Mateusz did [here](https://github.com/spencermountain/wikipedia-to-mongodb/pull/8), getting it to work in a queue - That's beyond my paygrade, but pretty cool.
ah, any changes to the layout, i welcome. go nuts.
Hey, aren't using grunt already?
cheers
-spence

**Comment by soyjavi** [comment]
Hey,

I think is not sync library 'cause you have a callback waiting for wikipedia response :) ... so, for code-style maybe is better use .then/catch (actually you dont have nothing for dispatch errors).

About grunt... 2/3 years ago I moved to Gulp... and last year I moved to Webpack... and now I combine webpack + npm scripts :)

best,

**Comment by rightisleft** [comment]
Error handling would be a great help - thats what promises are for. I can't seem to properly trap 404 errors on a request.

**Comment by spencermountain** [comment]
yeah, you guys are right.
I've got a busy couple weeks, but any sort of pr is welcomed.

**Comment by derhuerst** [comment]
@soyjavi I just found this library and am happy that there seems to be a reasonably focused npm module (in opposition to Parsoid) that converts Wikitext. are still following this?

**Comment by tothimre** [comment]
I am wondering if is there a reason you don't use give back node-like callbacks(err, article) style, because with the node8+ promisify magic could make it a promise for anyone who wants it.

**Comment by spencermountain** [comment]
Hey Imre, yeah, promises are in the new dev branch. I agree it's a good idea to support. I think what we're doing is returning a promise unless a callback is given as a param... So both workflows are supported. Will try to release this by next weekend, it's a big change.
Cheers

**Comment by spencermountain** [comment]
hey all! this is implemented in 3.0.0

```js
wtf.fetch(page) //returns a Promise
```
🎸

**Comment by niebert** [comment]
Spencer you are great, wonderful to have this library and other people seeing the value of your work. cheers, Bert
